### ISSUE TITLE
uri_string:parse/1 strips [] from IPv6 address literals in host field

### ISSUE THREAD

**Issue by dch** [erlang/otp#4731]
**Describe the bug**

An IPv6 literal is the equivalent of `127.0.0.1`, e.g. `[fe80::]`. Per [RFC2732](https://tools.ietf.org/html/rfc2732) hostnames in URIs should be represented with the `[]` included.

Thanks @polymetis for the paired code spelunking to track this down.

**To Reproduce**

```
1> uri_string:parse("http://user:password@[fe80::]:1337/?foo=bar").
#{host => "fe80::",  <------------ wrapping is missing
  path => "/",
  port => 1337,
  query => "foo=bar",
  scheme => "http",
  userinfo => "user:password"}

2> maps:get(host, uri_string:parse("http://user:password@[fe80::]:1337/?foo=bar")).
"fe80::"
```

**Expected behavior**

```
2> maps:get(host, uri_string:parse("http://user:password@[fe80::]:1337/?foo=bar")).
"[fe80::]"
```

**Affected versions**

* `Erlang/OTP 23 [erts-11.1.8] [source] [64-bit] [smp:32:32] [ds:32:32:10] [async-threads:1]`

**Additional context**

***References***

- [wikipedia](https://en.wikipedia.org/wiki/IPv6)  for a relatively digestable explanation

> Because IPv6 addresses contain colons, and HTTP uses colons as the port-address separator,
> RFC2732[39] specifies to use [square brackets] when using IPv6 addresses in URLs. e.g. > http://[2001:db8:4006:812::200e] (or, with a port, http://[2001:db8:4006:812::200e]:8080/path/page.html)

- normative ABNF  from [RFC3986 appendix A](https://tools.ietf.org/html/rfc3986#appendix-A) which updated and obsoleted RFC2732

```
      IP-literal = "[" ( IPv6address / IPvFuture  ) "]"
```

- https://github.com/elixir-lang/elixir/issues/10896

***Background***

When dealing with IPv6 only hostnames (servers that *only* have ip6 `AAAA` records in DNS), such as https://ipv6.google.com/ the impact is that downstream libraries try to treat the `IPv6 literal address` as a hostname, which obviously simply doesn't work. This fails in several places, for example, in `httpc` or `gen_tcp` when the `userinfo` part of the parsed URI map is removed (to be re-used as a header perhaps), and the ipv6 address is then used to reconstruct the URI from hostname.

***Options***

I can't tell how much downstream usage there is of the URI module, that depends on IPv6 literals being cosmetically pretty, but functionally broken. Thus:

- do nothing to avoid Breaking The Internet
- add a further field to the map that is functionally correct so that modules like httpc and gen_tcp can be downstream consumers of this new field

I'm happy to PR either case.

**Comment by peterdmv** [comment]
Some input to this issue. ```uri_map()``` is an intermediate data structure and thus it can have fields that are not RFC compliant by design.
```erlang
uri_map() =
    #{fragment => unicode:chardata(),
      host => unicode:chardata(),
      path => unicode:chardata(),
      port => integer() >= 0 | undefined,
      query => unicode:chardata(),
      scheme => unicode:chardata(),
      userinfo => unicode:chardata()}
```
URI consumers ```parse``` URIs into this intermediate structure and for some reason, we chose to not include syntactic clutter into this form e.g. brackets in IPv6 literals. When a library would like to use/create a URI it will become a URI producer and it should call ```recompose``` on the input ```uri_map()```. The output of ```recompose``` is an RFC compliant URI.

```erlang
8> uri_string:parse("http://user:password@[fe80::]:1337/?foo=bar"). 
#{host => "fe80::",path => "/",port => 1337,
  query => "foo=bar",scheme => "http",
  userinfo => "user:password"}
9> uri_string:recompose(v(8)).
"http://user:password@[fe80::]:1337/?foo=bar"
10> uri_string:recompose(#{host => "fe80::",path => "/",port => 1337}).
"//[fe80::]:1337/"
```

**Comment by josevalim** [comment]
That’s a reasonable argument. The other thing is that, if you are actually expecting an ipv6, having to remove the brackets would be annoying. So perhaps documenting it is enough. :)

**Comment by crertel** [comment]
What's the problem with following the RFC? The `host` field seems like something a reasonable person would expect to be formatted properly.

**Comment by rnowak** [comment]
> What's the problem with following the RFC? The `host` field seems like something a reasonable person would expect to be formatted properly.

Once it has been parsed and decomposed into its components, it no longer is a URI. I’d argue including the brackets wouldn’t be right as they are not part of the address.

**Comment by IngelaAndin** [comment]
httpc uses recompose, but still have some legacy code handling brackets, probably most of it is dead code except the code used to remove the brackets before calling gen_tcp that does not understand brackets that is an URI thing.

**Comment by lhoguin** [comment]
How do we know a host is IPv6? Does it always include `:` while others don't? Could this perhaps be documented?
